api/server/router/system/system_routes.go

```diff
@@ -57,12 +57,14 @@ func (s *systemRouter) swarmStatus() string {
 }
 
 func (s *systemRouter) getInfo(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
-	info := s.backend.SystemInfo()
-
-	if s.cluster != nil {
-		info.Swarm = s.cluster.Info()
-		info.Warnings = append(info.Warnings, info.Swarm.Warnings...)
-	}
+	info, _, _ := s.collectSystemInfo.Do(ctx, struct{}{}, func(ctx context.Context) (*types.Info, error) {
+		info := s.backend.SystemInfo()
+		if s.cluster != nil {
+			info.Swarm = s.cluster.Info()
+			info.Warnings = append(info.Warnings, info.Swarm.Warnings...)
+		}
+		return info, nil
+	})
 
 	version := httputils.VersionFromContext(ctx)
 	if versions.LessThan(version, "1.25") {
```

Comment:
Hm.. so thinking about this; if `info` (as returned above) is gonna be shared, and it's a pointer, then I guess the steps below wouldn't really work (as they would now modify all requests?

Reply:
Fixed this by using the API version as key for the single-flight